src/System.Private.ServiceModel/tests/Common/Infrastructure/src/ServiceUtilHelper.cs

```diff
@@ -11,70 +11,93 @@
 
 public static class ServiceUtilHelper
 {
-    private const string ClientCertificateSubject = "WCF Client Certificate";  
-    private const string CertificateIssuer = "DO_NOT_TRUST_WcfBridgeRootCA";  
+    private const string ClientCertificateSubject = "WCF Client Certificate";
+    private const string CertificateIssuer = "DO_NOT_TRUST_WcfBridgeRootCA";
 
     private static object s_certLock = new object();
     private static string s_serviceHostName = string.Empty;
     private static bool s_rootCertAvailabilityChecked = false;
     private static bool s_clientCertAvailabilityChecked = false;
+    private static X509Certificate2 s_rootCertificate = null;
+    private static X509Certificate2 s_clientCertificate = null;
+    private static string s_rootCertInstallErrorMessage = null;
+    private static string s_clientCertInstallErrorMessage = null;
 
-    public static X509Certificate2 RootCertificate { get; private set; }
+    public static X509Certificate2 RootCertificate
+    {
+        get
+        {
+            ThrowIfRootCertificateInstallationError();
+            return s_rootCertificate;
+        }
+    }
 
-    public static X509Certificate2 ClientCertificate { get; private set; }
+    public static X509Certificate2 ClientCertificate
+    {
+        get
+        {
+            ThrowIfClientCertificateInstallationError();
+            return s_clientCertificate;
+        }
+    }
 
     // Tries to ensure that the root certificate is installed into
     // the root store.  It obtains the root certificate from the service
     // utility endpoint and either installs it or verifies that a matching
-    // one is already installed.  A 'true' return indicates there is a root
-    // certificate in the store and available via 'RootCertificate'.
-    public static bool TryEnsureRootCertificateInstalled()
+    // one is already installed.  InvalidOperationException will be thrown
+    // if an error occurred attempting to install the certificate.  This
+    // method may be called multiple times but will attempt the installation
+    // once only.
+    public static void EnsureRootCertificateInstalled()
     {
-        lock (s_certLock)
+        if (!s_clientCertAvailabilityChecked)
```

Comment:
Why are we keying off *client*Cert here rather than *root*CertAvailabilityChecked?